torch/_inductor/fx_passes/joint_graph.py

```diff
@@ -333,3 +344,119 @@ def pointless_view(match: Match, arg, size):
     if size == arg_size:
         node.replace_all_uses_with(node.args[0])
         match.erase_nodes(graph)
+
+
+# When softmax is used with temperature or other scaling, we get the pattern
+#
+#   scale(x) - scale(x).amax(dim, keepdim=True)
+#
+# which is expected to be at most zero, but we may end up with numerical
+# discrepancies # between the recomputed values of scale(x) inside and out
+# of the reduction, # depending on compiler optimizations, e.g. use of fma
+# instructions.
+#
+# Here we replace it with the mathematically equivalent,
+#
+#   scale(x - x.amax(dim, keepdim=True))
+#
+# which is more stable as we only compute the scaling once.
+#
+# NOTE: This pattern must come after fused attention matching!
```

Comment:
Can we test this?

Reply:
I think the existing fused attention tests are sufficient. If this pattern is applied first then the attention patterns wouldn't match any more.